cornell-dti/office-hours
# Completely new version of comments

### Summary 
This might be one of my most ambitious PRs ever. This adds the new comments feature for TAs, students, and professors, which completely changes how comments are run.

This PR allows students, TAs, and professors to comment in a thread-like style with one another, facilitating better communication. The specifics of the new features that were added are detailed below.



- Created a completely new type for comments, FireComment, and structured it so that each question contains a comments collection once comments are made
- New reply button, as well as new look for no show, done, assign from me/unassign from me, and remove buttons
- Create a new comment by clicking on the reply button, entering your comment, then clicking post
- Edit and delete comments through use of the buttons (delete is hidden by default and is accessed by clicking on the three dots ... in the comment, as it is a destructive action)
- Edited security rules to allow only TAs/professors/the student asking the question to read the comment and to allow only the user who posted the comment to delete or edit the comment
- Click on the gray bar to the side to collapse comments (i.e. hide them)
- On the TA/professor side, allow TAs/professors to see the names and profile pictures of commenters, and on the student side, allow them to see if a TA posted a comment or if it is their own comment
- Sorted comments so that the most recent comments come up first




### Test Plan 
- Try adding/editing/deleting a comment, both from the professor/TA side and the user side
- Check the Firebase security rules and test them to see if they work
- Check to see if the user profiles/names show up for TAs and just an anonymous profile picture and generalized name (TA or student) for students
- Make sure the no show, done, assign from me, unassign from me, and remove buttons still work


### Notes 



### Breaking Changes 

Firebase security policy change
I updated existing types in `/src/components/types/`
